Implement load generator for cockroachdb

gprasath opened the pull request:
Signed-off-by: gprasath <giridhara.prasad@cloudbyte.com>

**What this PR does / why we need it**:
- Implemented workload generator for cockroachdb application.
- Generates load on cockroachdb using KV.
- This is implemented considering that cockroachdb application is already deployed in the cluster.

gprasath commented:
```giridhara_prasad@cloudshell:~/litmus/apps/cockroachdb/workload (strong-eon-153112)$ kubectl logs cockroachdb-loadgen-rl55z -n litmus
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "f0ceb61d902ca0ed86b3ab7eba74127846bacf1d", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "d4e99cb22a812c9f6dc36a6c8f8b9e51", "mode": "0644", "owner": "root", "size": 428, "src": "/root/.ansible/tmp/ansible-tmp-1537522682.23-186080936635224/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.331451", "end": "2018-09-21 09:38:04.440569", "failed_when_result": false, "rc": 0, "start": "2018-09-21 09:38:03.109118", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-load-generation\" created", "stdout_lines": ["litmusresult \"cockroachdb-load-generation\" created"]}

TASK [Creating cockroachdb loadgen job] ****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f cockroachdb_loadgen.yml -n cockroachdb", "delta": "0:00:01.072690", "end": "2018-09-21 09:38:05.738718", "failed_when_result": false, "rc": 0, "start": "2018-09-21 09:38:04.666028", "stderr": "", "stderr_lines": [], "stdout": "job \"cockroachdb-lg\" created", "stdout_lines": ["job \"cockroachdb-lg\" created"]}

TASK [Verifying whether the cockroachdb loadgen pod is running.] ***************
FAILED - RETRYING: Verifying whether the cockroachdb loadgen pod is running. (15 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl get pods -n cockroachdb -l app=cockroachdb-lg", "delta": "0:00:01.373768", "end": "2018-09-21 09:38:38.688911", "rc": 0, "start": "2018-09-21 09:38:37.315143", "stderr": "", "stderr_lines": [], "stdout": "NAME                   READY     STATUS    RESTARTS   AGE\ncockroachdb-lg-8hslz   1/1       Running   0          33s", "stdout_lines": ["NAME                   READY     STATUS    RESTARTS   AGE", "cockroachdb-lg-8hslz   1/1       Running   0          33s"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "loadgen_started"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "20049c64143f8af921082b3616ffcae29fd8f1d8", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "31d1d8075ae7d397107a2f7ffef71ba1", "mode": "0644", "owner": "root", "size": 437, "src": "/root/.ansible/tmp/ansible-tmp-1537522718.93-264927772133075/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.755716", "end": "2018-09-21 09:38:43.684057", "failed_when_result": false, "rc": 0, "start": "2018-09-21 09:38:41.928341", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-load-generation\" configured", "stdout_lines": ["litmusresult \"cockroachdb-load-generation\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=8    changed=6    unreachable=0    failed=0```

ksatchit commented on apps/cockroachdb/workload/test.yml:
```diff
@@ -0,0 +1,79 @@
+          register: result
+          failed_when: 'result.rc != 0'
+
+        - name: Verifying whether the cockroachdb loadgen pod is running.
```
Would be great to have a check that loadgen has 'really' written data

gprasath replied:
Thanks @ksatchit . This creates a database named 'test' with table 'kv' loaded with randomly generated key,values. Now, checking if such table resides inside pod.

gprasath commented:
```
TASK [Obtain the cockroachdb pod name] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pods -n cockroachdb -l app=cockroachdb -o jsonpath='{.items[0].metadata.name}'", "delta": "0:00:01.959874", "end": "2018-09-24 13:59:43.776198", "rc": 0, "start": "2018-09-24 13:59:41.816324", "stderr": "", "stderr_lines": [], "stdout": "cockroachdb-0", "stdout_lines": ["cockroachdb-0"]}

TASK [Verify the data written] *************************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl exec -it cockroachdb-0 -n cockroachdb -- /cockroach/cockroach sql --insecure --host=cockroachdb-public.cockroachdb -d test -e 'show tables;'", "delta": "0:00:01.430544", "end": "2018-09-24 13:59:45.499205", "rc": 0, "start": "2018-09-24 13:59:44.068661", "stderr": "Unable to use a TTY - input is not a terminal or the right kind of file", "stderr_lines": ["Unable to use a TTY - input is not a terminal or the right kind of file"], "stdout": "Table\nkv\n# 1 row", "stdout_lines": ["Table", "kv", "# 1 row"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "a66d9c4a7873a04c6769023d8b07b6c2acd6fac4", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "addcb05b49da52e849286f42b554b6fd", "mode": "0644", "owner": "root", "size": 426, "src": "/root/.ansible/tmp/ansible-tmp-1537797585.75-127588678039765/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.335396", "end": "2018-09-24 13:59:49.767201", "failed_when_result": false, "rc": 0, "start": "2018-09-24 13:59:48.431805", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-load-generation\" configured", "stdout_lines": ["litmusresult \"cockroachdb-load-generation\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=9    changed=7    unreachable=0    failed=0```

gprasath commented:
Using label for loadgen pod:
```giridhara_prasad@cloudshell:~/litmus/apps/cockroachdb/workload (strong-eon-153112)$ kubectl get po -n litmus --show-labels
NAME                        READY     STATUS      RESTARTS   AGE       LABELS
cockroachdb-loadgen-fhd7q   0/1       Completed   0          33s       controller-uid=a02d45de-c0b0-11e8-8e31-42010a8000f5,job-name=cockroachdb-loadgen,loadgen=cockroachdb-load```

gprasath commented:
- Passing time duration, namespace and service name as env.
- Replacing above variables in loadgen.yml

```
giridhara_prasad@cloudshell:~/litmus (strong-eon-153112)$ kubectl logs cockroachdb-loadgen-82zjm -n litmus
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "f0ceb61d902ca0ed86b3ab7eba74127846bacf1d", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "d4e99cb22a812c9f6dc36a6c8f8b9e51", "mode": "0644", "owner": "root", "size": 428, "src": "/root/.ansible/tmp/ansible-tmp-1537940093.04-55081908185026/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.338974", "end": "2018-09-26 05:34:55.217279", "failed_when_result": false, "rc": 0, "start": "2018-09-26 05:34:53.878305", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-load-generation\" configured", "stdout_lines": ["litmusresult \"cockroachdb-load-generation\" configured"]}

TASK [Replace the arguments with values in loadgen yaml] ***********************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Replace the service name in loadgen yaml] ********************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Update the time interval in loadgen yaml] ********************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Creating cockroachdb loadgen job] ****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f cockroachdb_loadgen.yml -n cockroachdb", "delta": "0:00:01.067200", "end": "2018-09-26 05:34:57.249760", "failed_when_result": false, "rc": 0, "start": "2018-09-26 05:34:56.182560", "stderr": "", "stderr_lines": [], "stdout": "job \"cockroachdb-lg\" created", "stdout_lines": ["job \"cockroachdb-lg\" created"]}

TASK [Obtain the cockroachdb pod name] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pods -n cockroachdb -l app=cockroachdb -o jsonpath='{.items[0].metadata.name}'", "delta": "0:00:01.015227", "end": "2018-09-26 05:34:58.503272", "rc": 0, "start": "2018-09-26 05:34:57.488045", "stderr": "", "stderr_lines": [], "stdout": "cockroachdb-0", "stdout_lines": ["cockroachdb-0"]}

TASK [Verify the data written] *************************************************
FAILED - RETRYING: Verify the data written (5 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl exec -it cockroachdb-0 -n cockroachdb -- /cockroach/cockroach sql --insecure --host=cockroachdb-public.cockroachdb -d test -e 'show tables;'", "delta": "0:00:01.412295", "end": "2018-09-26 05:36:01.478279", "rc": 0, "start": "2018-09-26 05:36:00.065984", "stderr": "Unable to use a TTY - input is not a terminal or the right kind of file", "stderr_lines": ["Unable to use a TTY - input is not a terminal or the right kind of file"], "stdout": "Table\nkv\n# 1 row", "stdout_lines": ["Table", "kv", "# 1 row"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "a66d9c4a7873a04c6769023d8b07b6c2acd6fac4", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "addcb05b49da52e849286f42b554b6fd", "mode": "0644", "owner": "root", "size": 426, "src": "/root/.ansible/tmp/ansible-tmp-1537940161.68-27362422425460/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.277006", "end": "2018-09-26 05:36:05.416945", "failed_when_result": false, "rc": 0, "start": "2018-09-26 05:36:04.139939", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-load-generation\" configured", "stdout_lines": ["litmusresult \"cockroachdb-load-generation\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=12   changed=10   unreachable=0    failed=0
```

gprasath commented:
- Obtaining application label and loadgen pod label as env.
- Obtaining service name using app label and namespace
- Forming FQDN and updating it in loadgen spec.
- Updating loadgen label in loadgen spec.


```
giridhara_prasad@cloudshell:~/litmus/apps/cockroachdb/workload (strong-eon-153112)$ kubectl logs -f cockroachdb-loadgen-8t9t4 -n litmus
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "f0ceb61d902ca0ed86b3ab7eba74127846bacf1d", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "d4e99cb22a812c9f6dc36a6c8f8b9e51", "mode": "0644", "owner": "root", "size": 428, "src": "/root/.ansible/tmp/ansible-tmp-1538029760.28-40464084036848/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.272615", "end": "2018-09-27 06:29:22.376816", "failed_when_result": false, "rc": 0, "start": "2018-09-27 06:29:21.104201", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-load-generation\" configured", "stdout_lines": ["litmusresult \"cockroachdb-load-generation\" configured"]}

TASK [Obtaining the application label from env.] *******************************
ok: [127.0.0.1] => {"ansible_facts": {"app_lkey": "app", "app_lvalue": "cockroachdb"}, "changed": false}

TASK [Obtaining the service name.] *********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get svc -n cockroachdb -l app=cockroachdb -o jsonpath='{.items[0].metadata.name}'", "delta": "0:00:00.904223", "end": "2018-09-27 06:29:23.557643", "rc": 0, "start": "2018-09-27 06:29:22.653420", "stderr": "", "stderr_lines": [], "stdout": "cockroachdb", "stdout_lines": ["cockroachdb"]}

TASK [Recording the service name.] *********************************************
ok: [127.0.0.1] => {"ansible_facts": {"svc_name": "cockroachdb"}, "changed": false}

TASK [Form FQDN using namespace and service name.] *****************************
ok: [127.0.0.1] => {"ansible_facts": {"fqdn": "cockroachdb.cockroachdb.svc.cluster.local"}, "changed": false}

TASK [Obtaining the loadgen pod label from env.] *******************************
ok: [127.0.0.1] => {"ansible_facts": {"loadgen_lkey": "loadgen", "loadgen_lvalue": "cockroachdb-load"}, "changed": false}

TASK [Replace the label in loadgen job spec.] **********************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Place the FQDN in loadgen yaml.] *****************************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Updating the time duration in loadgen yaml.] *****************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Creating cockroachdb loadgen job.] ***************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f cockroachdb_loadgen.yml -n cockroachdb", "delta": "0:00:01.023622", "end": "2018-09-27 06:29:25.795335", "failed_when_result": false, "rc": 0, "start": "2018-09-27 06:29:24.771713", "stderr": "", "stderr_lines": [], "stdout": "job \"cockroachdb-lg\" created", "stdout_lines": ["job \"cockroachdb-lg\" created"]}

TASK [Obtain the cockroachdb pod name] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pods -n cockroachdb -l app=cockroachdb -o jsonpath='{.items[0].metadata.name}'", "delta": "0:00:01.189087", "end": "2018-09-27 06:29:27.207500", "rc": 0, "start": "2018-09-27 06:29:26.018413", "stderr": "", "stderr_lines": [], "stdout": "cockroachdb-0", "stdout_lines": ["cockroachdb-0"]}

TASK [Verify the data written] *************************************************
FAILED - RETRYING: Verify the data written (5 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl exec -it cockroachdb-0 -n cockroachdb -- /cockroach/cockroach sql --insecure --host=cockroachdb-public.cockroachdb -d test -e 'show tables;'", "delta": "0:00:01.438209", "end": "2018-09-27 06:30:30.400088", "rc": 0, "start": "2018-09-27 06:30:28.961879", "stderr": "Unable to use a TTY - input is not a terminal or the right kind of file", "stderr_lines": ["Unable to use a TTY - input is not a terminal or the right kind of file"], "stdout": "Table\nkv\n# 1 row", "stdout_lines": ["Table", "kv", "# 1 row"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "a66d9c4a7873a04c6769023d8b07b6c2acd6fac4", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "addcb05b49da52e849286f42b554b6fd", "mode": "0644", "owner": "root", "size": 426, "src": "/root/.ansible/tmp/ansible-tmp-1538029830.63-170646094268038/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.387165", "end": "2018-09-27 06:30:34.577045", "failed_when_result": false, "rc": 0, "start": "2018-09-27 06:30:33.189880", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"cockroachdb-load-generation\" configured", "stdout_lines": ["litmusresult \"cockroachdb-load-generation\" configured"]}

PLAY RECAP *********************************************************************
```

ksatchit commented on apps/cockroachdb/workload/test.yml:
```diff
@@ -0,0 +1,130 @@
+        - name: Verify the data written
+          shell: >
+            kubectl exec -it {{ pod_name.stdout }} -n {{ namespace }} -- /cockroach/cockroach sql --insecure 
+            --host=cockroachdb-public.cockroachdb -d test -e 'show tables;' 
```
@gprasath , is the hostname specific to the image of cockroachdb we are using ?
